File: apps/mobile/utils/chat-events.ts
import type { AgentMessage, AgentSessionEvent, ToolResult } from "@console/types";
import type {
  ActivityEvent,
  ChatSessionState,
  ChatSnapshot,
  RunActivityState,
} from "@/types/chat";

/** Unique per-message id (crypto.randomUUID with a fallback). */
export function newMessageId(): string {
  if (typeof crypto !== "undefined" && "randomUUID" in crypto) {
    return crypto.randomUUID();
  }
  return `msg-${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

/**
 * Ensure every message carries a unique id.
 *
 * Legend State keys observable arrays by `id` and warns when elements collide;
 * assistant messages get ids from the server but user/toolResult messages are
 * created without one. Returns the same array reference when nothing changed
 * so callers don't churn object identities unnecessarily.
 */
export function ensureMessageIds(messages: AgentMessage[]): AgentMessage[] {
  let changed = false;
  const out = messages.map((m) => {
    if (m.id) return m;
    changed = true;
    return { ...m, id: newMessageId() };
  });
  return changed ? out : messages;
}

/** Update the latest run in the session's runs array. */
function updateLatestRun(
  session: ChatSessionState,
  update: (run: RunActivityState) => RunActivityState,
): ChatSessionState {
  if (session.runs.length === 0) return session;
  const runs = [...session.runs];
  runs[runs.length - 1] = update(runs[runs.length - 1]!);
  return { ...session, runs };
}

/** Update a specific tool call event's result by toolCallId. */
function setToolCallResult(events: ActivityEvent[], result: ToolResult): ActivityEvent[] {
  return events.map((event) =>
    event.type === "toolCall" && event.call.id === result.toolCallId
      ? { ...event, result }
      : event,
  );
}

/** Finalize any tool call events that never received a result. */
function finalizePendingToolCalls(events: ActivityEvent[]): ActivityEvent[] {
  return events.map((event) =>
    event.type === "toolCall" && !event.result
      ? {
          ...event,
          result: {
            toolCallId: event.call.id,
            toolName: event.call.name,
            content: "Run ended before this tool call completed.",
            isError: true,
          },
        }
      : event,
  );
}

export function applyChatEvent(
  session: ChatSessionState,
  event: AgentSessionEvent,
): ChatSessionState {
  switch (event.type) {
    case "modelStreamPart": {
      const text = event.part?.text;
      const thinking = event.part?.thinking;
      if (!text && !thinking) return session;
      return {
        ...session,
        streamingText: text ? session.streamingText + text : session.streamingText,
        streamingThinking: thinking
          ? session.streamingThinking + thinking
          : session.streamingThinking,
      };
    }
    case "modelStreamEnd": {
      const turn = event.turn;
      if (!turn) {
        // Fallback: construct from streaming buffers if no turn was provided.
        if (!session.streamingText && !session.streamingThinking) return session;
        return {
          ...session,
          messages: [
            ...session.messages,
            {
              role: "assistant",
              id: newMessageId(),
              createdAt: Date.now(),
              content: [
                ...(session.streamingThinking
                  ? [{ type: "thinking" as const, text: session.streamingThinking }]
                  : []),
                ...(session.streamingText
                  ? [{ type: "text" as const, text: session.streamingText }]
                  : []),
              ],
            },
          ],
          streamingText: "",
          streamingThinking: "",
        };
      }

      const turnWithMeta = turn as typeof turn & { createdAt?: number; id?: string };
      const normalizedTurn = {
        ...turnWithMeta,
        // Turns may arrive without an id (provider-dependent) — assign one so
        // every stored message is uniquely keyed.
        id: turnWithMeta.id ?? newMessageId(),
        createdAt: turnWithMeta.createdAt ?? Date.now(),
      };

      // Always append the turn to messages for persistence.
      const baseResult: ChatSessionState = {
        ...session,
        messages: [...session.messages, normalizedTurn],
        streamingText: "",
        streamingThinking: "",
      };

      // If the turn has tool calls, extract text and tool call parts as
      // timeline events in the latest run. The text becomes "progress text"
      // inside the run activity, not a standalone message bubble.
      const toolCallParts = turn.content.filter(
        (c): c is Extract<typeof c, { type: "toolCall" }> => c.type === "toolCall",
      );

      if (toolCallParts.length === 0) {
        // No tool calls — this is the final response, not part of the timeline.
        return baseResult;
      }

      // Build timeline events in content order: thinking/text parts become
      // thinking/text events, tool call parts become toolCall events (with no
      // result yet). This preserves the chronological think → tool → think →
      // tool order from the agent loop.
      const newEvents: ActivityEvent[] = [];
      for (const part of turn.content) {
        if (part.type === "thinking" && part.text.trim()) {
          newEvents.push({
            type: "thinking",
            id: `thinking-${part.text.slice(0, 16)}-${Date.now()}`,
            text: part.text,
          });
        } else if (part.type === "text" && part.text.trim()) {
          newEvents.push({
            type: "text",
            id: `text-${part.text.slice(0, 16)}-${Date.now()}`,
            text: part.text,
          });
        } else if (part.type === "toolCall") {
          newEvents.push({ type: "toolCall", id: part.call.id, call: part.call });
        }
      }

      // Set active tool calls for toolExecutionEnd error finalization.
      baseResult.activeToolCalls = toolCallParts.map((c) => c.call);

      return updateLatestRun(baseResult, (run) => ({
        ...run,
        events: [...run.events, ...newEvents],
      }));
    }
    case "toolExecutionResult":
      return updateLatestRun(session, (run) => ({
        ...run,
        events: setToolCallResult(run.events, event.result),
      }));
    case "toolExecutionStart": {
      // Tool calls were already added as events in modelStreamEnd.
      // Just track active calls for toolExecutionEnd finalization.
      return { ...session, activeToolCalls: event.calls };
    }
    case "toolExecutionEnd": {
      // Build the complete results list for this batch.
      const eventResults = [...event.results];
      // Add error results for any active calls that didn't get a result.
      for (const call of session.activeToolCalls) {
        if (!eventResults.some((r) => r.toolCallId === call.id)) {
          eventResults.push({
            toolCallId: call.id,
            toolName: call.name,
            content: "Tool execution ended without a result.",
            isError: true,
          });
        }
      }

      // Update each tool call event with its result.
      let updated = updateLatestRun(session, (run) => ({
        ...run,
        events: eventResults.reduce(setToolCallResult, run.events),
      }));

      // Append the toolResult message to messages for persistence transport.
      // It renders as null in the UI — results are shown via RunActivity.
      updated = {
        ...updated,
        messages: [
          ...updated.messages,
          { role: "toolResult", id: newMessageId(), results: eventResults },
        ],
        activeToolCalls: [],
      };

      return updated;
    }
    case "askQuestion":
      return {
        ...session,
        pendingQuestions: [...session.pendingQuestions, { request: event.request }],
      };
    case "permissionRequest":
      return {
        ...session,
        pendingPermissions: [...session.pendingPermissions, { request: event.request }],
      };
    case "todoUpdate":
      return { ...session, todoItems: event.items };
    case "sessionEnd":
      // Finalize the latest run: mark as completed if still working, and
      // add error results for any tool calls that never received a result.
      return updateLatestRun(session, (run) => {
        if (run.status !== "working") return run;
        return {
          ...run,
          status: "completed",
          events: finalizePendingToolCalls(run.events),
          elapsedMs: run.startedAt ? Date.now() - run.startedAt : run.elapsedMs,
        };
      });
    case "subagentStart": {
      const subagents = session.subagents ? [...session.subagents] : [];
      const existingIdx = subagents.findIndex((s) => s.subagentId === event.subagentId);
      if (existingIdx >= 0) {
        subagents[existingIdx] = {
          ...subagents[existingIdx]!,
          name: event.name,
          role: event.role,
          prompt: event.prompt,
          maxTurns: event.maxTurns,
          status: "running",
          updatedAt: Date.now(),
        };
      } else {
        subagents.push({
          subagentId: event.subagentId,
          parentToolCallId: event.parentToolCallId,
          name: event.name,
          role: event.role,
          prompt: event.prompt,
          maxTurns: event.maxTurns,
          currentTurn: 1,
          status: "running",
          activities: [],
          createdAt: Date.now(),
          updatedAt: Date.now(),
        });
      }
      return {
        ...session,
        subagents,
      };
    }
    case "subagentActivity": {
      const subagents = session.subagents ? [...session.subagents] : [];
      const idx = subagents.findIndex((s) => s.subagentId === event.subagentId);
      if (idx >= 0) {
        const target = { ...subagents[idx]! };
        target.currentTurn = Math.max(target.currentTurn, event.turnIndex);
        const activities = [...target.activities];
        const actIdx = activities.findIndex((a) => a.toolCallId === event.toolCallId);
        const existing = actIdx >= 0 ? activities[actIdx] : undefined;
        const item = {
          turnIndex: event.turnIndex,
          toolCallId: event.toolCallId,
          toolName: event.toolName || existing?.toolName || "",
          args: event.args ?? existing?.args,
          status: event.status,
          error: event.error ?? existing?.error,
        };
        if (actIdx >= 0) {
          activities[actIdx] = item;
        } else {
          activities.push(item);
        }
        target.activities = activities;
        target.updatedAt = Date.now();
        subagents[idx] = target;
      }
      return {
        ...session,
        subagents,
      };
    }
    case "subagentEnd": {
      const subagents = session.subagents ? [...session.subagents] : [];
      const idx = subagents.findIndex((s) => s.subagentId === event.subagentId);
      if (idx >= 0) {
        const target = { ...subagents[idx]! };
        target.status = event.status;
        if (event.summary !== undefined) target.summary = event.summary;
        if (event.error !== undefined) target.error = event.error;
        if (event.totalTurns > 0) target.currentTurn = event.totalTurns;
        target.updatedAt = Date.now();
        subagents[idx] = target;
      }
      return {
        ...session,
        subagents,
      };
    }
    case "streamReset":
      // Re-attach replay is about to arrive — clear streaming buffers so the
      // replayed coalesced snapshots don't duplicate already-seen text.
      return {
        ...session,
        streamingText: "",
        streamingThinking: "",
      };
    case "error":
      if (event.error?.message.toLowerCase().includes("aborted")) {
        return updateLatestRun(session, (run) => ({
          ...run,
          status: run.status === "working" ? "aborted" : run.status,
          events: finalizePendingToolCalls(run.events),
        }));
      }
      return {
        ...updateLatestRun(session, (run) => ({
          ...run,
          status: run.status === "working" ? "failed" : run.status,
          events: finalizePendingToolCalls(run.events),
        })),
        messages: [
          ...session.messages,
          {
            role: "assistant",
            id: newMessageId(),
            content: [
              { type: "text", text: `Error: ${event.error?.message ?? "Unknown agent error"}` },
            ],
          },
        ],
        streamingText: "",
        streamingThinking: "",
      };
    default:
      return session;
  }
}

/**
 * Derive the UI snapshot for a session.
 *
 * The store holds the full per-session runtime state (queues, runs timeline,
 * todo items, attachments). The snapshot flattens it to what screens render:
 * a single pending permission/question (the most recent), live tool results
 * from the latest run, and the streaming buffers.
 *
 * Snapshots are cached per session-state reference so selectors handed to
 * `useSyncExternalStore` return a stable object identity for unchanged state.
 * Without this, React 19 reports "The result of getSnapshot should be cached"
 * and can loop with "Maximum update depth exceeded".
 */
const snapshotCache = new WeakMap<ChatSessionState, ChatSnapshot>();

export function toChatSnapshot(session: ChatSessionState): ChatSnapshot {
  const cached = snapshotCache.get(session);
  if (cached) return cached;

  const snapshot: ChatSnapshot = {
    messages: session.messages,
    streamingText: session.streamingText,
    streamingThinking: session.streamingThinking,
    activeToolCalls: session.activeToolCalls,
    liveToolResults: (session.runs[session.runs.length - 1]?.events ?? [])
      .filter(
        (event): event is Extract<ActivityEvent, { type: "toolCall" }> =>
          event.type === "toolCall" && Boolean(event.result),
      )
      .map((event) => event.result!),
    pendingPermission: session.pendingPermissions[session.pendingPermissions.length - 1] ?? null,
    pendingQuestion: session.pendingQuestions[session.pendingQuestions.length - 1] ?? null,
    pendingPermissions: session.pendingPermissions,
    pendingQuestions: session.pendingQuestions,
    todoItems: session.todoItems ?? [],
    subagents: session.subagents ?? [],
    running: session.running,
    runs: session.runs,
  };

  snapshotCache.set(session, snapshot);
  return snapshot;
}

File: apps/android/app/src/main/kotlin/com/console/mobile/core/chat/ChatEvents.kt
package com.console.mobile.core.chat

import com.console.mobile.data.model.AgentMessage
import com.console.mobile.data.model.AgentSessionEvent
import com.console.mobile.data.model.AssistantMessage
import com.console.mobile.data.model.ThinkingPart
import com.console.mobile.data.model.TextPart
import com.console.mobile.data.model.ToolCallPart
import com.console.mobile.data.model.ToolResult
import com.console.mobile.data.model.UserMessage
import com.console.mobile.data.model.ToolResultMessage
import java.util.UUID

fun newMessageId(): String = try {
    UUID.randomUUID().toString()
} catch (_: Exception) {
    "msg-${System.currentTimeMillis()}-${(Math.random() * 1e6).toInt()}"
}

fun ensureMessageIds(messages: List<AgentMessage>): List<AgentMessage> {
    var changed = false
    val out = messages.map { m ->
        if (m.id != null) m else {
            changed = true
            when (m) {
                is UserMessage -> m.copy(id = newMessageId())
                is AssistantMessage -> m.copy(id = newMessageId())
                is ToolResultMessage -> m.copy(id = newMessageId())
            }
        }
    }
    return if (changed) out else messages
}

private fun updateLatestRun(
    session: ChatSessionState,
    update: (RunActivityState) -> RunActivityState,
): ChatSessionState {
    if (session.runs.isEmpty()) return session
    val runs = session.runs.toMutableList()
    runs[runs.lastIndex] = update(runs.last())
    return session.copy(runs = runs)
}

private fun setToolCallResult(events: List<ActivityEvent>, result: ToolResult): List<ActivityEvent> =
    events.map { e ->
        if (e is ActivityEvent.ToolCallEvent && e.call.id == result.toolCallId) e.copy(result = result) else e
    }

private fun finalizePendingToolCalls(events: List<ActivityEvent>): List<ActivityEvent> =
    events.map { e ->
        if (e is ActivityEvent.ToolCallEvent && e.result == null) {
            e.copy(result = ToolResult(toolCallId = e.call.id, toolName = e.call.name, content = null, isError = true))
        } else e
    }

fun applyChatEvent(session: ChatSessionState, event: AgentSessionEvent): ChatSessionState {
    return when (event.type) {
        "modelStreamPart" -> {
            val text = event.part?.text.orEmpty()
            val thinking = event.part?.thinking.orEmpty()
            if (text.isEmpty() && thinking.isEmpty()) session
            else session.copy(
                streamingText = session.streamingText + text,
                streamingThinking = session.streamingThinking + thinking,
            )
        }
        "modelStreamEnd" -> {
            val turn = event.turn ?: run {
                if (session.streamingText.isEmpty() && session.streamingThinking.isEmpty()) return session
                val parts = buildList {
                    if (session.streamingThinking.isNotEmpty()) add(ThinkingPart(session.streamingThinking))
                    if (session.streamingText.isNotEmpty()) add(TextPart(session.streamingText))
                }
                return session.copy(
                    messages = session.messages + AssistantMessage(id = newMessageId(), createdAt = System.currentTimeMillis(), content = parts),
                    streamingText = "", streamingThinking = "",
                )
            }
            val normalized = turn.copy(id = turn.id ?: newMessageId(), createdAt = turn.createdAt ?: System.currentTimeMillis())
            var base = session.copy(messages = session.messages + normalized, streamingText = "", streamingThinking = "")
            val toolCalls = turn.content.filterIsInstance<ToolCallPart>()
            if (toolCalls.isEmpty()) return base
            val newEvents = mutableListOf<ActivityEvent>()
            for (part in turn.content) {
                when (part) {
                    is ThinkingPart -> if (part.text.isNotBlank()) newEvents.add(ActivityEvent.Thinking("thinking-${part.text.take(16)}-${System.currentTimeMillis()}", part.text))
                    is TextPart -> if (part.text.isNotBlank()) newEvents.add(ActivityEvent.Text("text-${part.text.take(16)}-${System.currentTimeMillis()}", part.text))
                    is ToolCallPart -> newEvents.add(ActivityEvent.ToolCallEvent(part.call.id, part.call))
                    else -> {}
                }
            }
            base = base.copy(activeToolCalls = toolCalls.map { it.call })
            updateLatestRun(base) { run -> run.copy(events = run.events + newEvents) }
        }
        "toolExecutionStart" -> session.copy(activeToolCalls = event.calls ?: emptyList())
        "toolExecutionResult" -> {
            val r = event.result ?: return session
            updateLatestRun(session) { run -> run.copy(events = setToolCallResult(run.events, r)) }
        }
        "toolExecutionEnd" -> {
            val results = event.results ?: emptyList()
            var s = session
            for (r in results) {
                s = updateLatestRun(s) { run -> run.copy(events = setToolCallResult(run.events, r)) }
            }
            s = updateLatestRun(s) { run ->
                run.copy(status = if (run.status == RunStatus.Working) RunStatus.Completed else run.status, events = finalizePendingToolCalls(run.events))
            }
            s.copy(activeToolCalls = emptyList())
        }
        "turnStart" -> session.copy(
            running = true,
            runs = session.runs + RunActivityState(runId = "run-${System.currentTimeMillis()}", startedAt = System.currentTimeMillis(), status = RunStatus.Working),
        )
        "turnEnd", "sessionEnd" -> {
            val finalized = updateLatestRun(session) { run ->
                if (run.status != RunStatus.Working) run
                else run.copy(status = RunStatus.Completed, events = finalizePendingToolCalls(run.events),
                    elapsedMs = if (run.startedAt != null) System.currentTimeMillis() - run.startedAt else run.elapsedMs)
            }
            finalized.copy(running = false, streamingText = "", streamingThinking = "", activeToolCalls = emptyList())
        }
        "permissionRequest" -> session // structured payload handled at store layer; keep reducer total
        "askQuestion" -> session
        "todoUpdate" -> session.copy(todoItems = event.items?.map { com.console.mobile.data.model.TodoItem(it.id, it.content, it.status) } ?: emptyList())
        "streamReset" -> session.copy(streamingText = "", streamingThinking = "", activeToolCalls = emptyList())
        "error" -> {
            val msg = event.error?.message ?: "Unknown agent error"
            if (msg.lowercase().contains("abort")) {
                updateLatestRun(session) { run ->
                    run.copy(status = if (run.status == RunStatus.Working) RunStatus.Aborted else run.status, events = finalizePendingToolCalls(run.events))
                }
            } else {
                val withRun = updateLatestRun(session) { run ->
                    run.copy(status = if (run.status == RunStatus.Working) RunStatus.Failed else run.status, events = finalizePendingToolCalls(run.events))
                }
                withRun.copy(
                    messages = withRun.messages + AssistantMessage(id = newMessageId(), content = listOf(TextPart("Error: $msg"))),
                    streamingText = "", streamingThinking = "",
                )
            }
        }
        else -> session
    }
}

fun toChatSnapshot(session: ChatSessionState): ChatSnapshot {
    val lastRun = session.runs.lastOrNull()
    val live = lastRun?.events?.filterIsInstance<ActivityEvent.ToolCallEvent>()?.mapNotNull { it.result } ?: emptyList()
    return ChatSnapshot(
        messages = session.messages,
        streamingText = session.streamingText,
        streamingThinking = session.streamingThinking,
        activeToolCalls = session.activeToolCalls,
        liveToolResults = live,
        pendingPermission = session.pendingPermissions.lastOrNull(),
        pendingQuestion = session.pendingQuestions.lastOrNull(),
        pendingPermissions = session.pendingPermissions,
        pendingQuestions = session.pendingQuestions,
        todoItems = session.todoItems,
        subagents = session.subagents,
        running = session.running,
        runs = session.runs,
    )
}
